### PR TITLE
Empty HashMap should work correctly without an insert.

### PR DESCRIPTION
**Pull request by morazow**
Hey @jonhoo,

Thanks for the great content!

I am new to Rust, while implementing simple HashMap from Youtube video, I have come across this issue. I have added a fix for it. 

Let me know if you want me to change anything.

### CODE REVIEW

**Comment by jonhoo** [comment]
Hey there!

Hmm, I _think_ the existing code should already work for this case. Does your test not pass without the other changes? What error do you get?

**Comment by morazow** [comment]
Yes, I tried to explain it in the commit message.

The error is division by zero because of taking modulo by [`buckets.len()`](https://github.com/jonhoo/rust-basic-hashmap/blob/master/src/lib.rs#L88) (no elements) when calculating the bucket index.

The panic error is `'attempt to calculate the remainder with a divisor of zero'`.

**Comment by jonhoo** [comment]
Ahh, sorry, I missed your commit message entirely! That's a good catch.

In that case, I still don't think you need to change `contains_key`, since it already calls `get`. So if `get` is fixed, `contains_key` is also indirectly fixed.

I also wonder if it'd be nicer for the `if` to only be in `fn bucket`. Change `fn bucket` to return an `Option`, and then just use `let bucket = self.bucket(key)?;` in `get` and `remove`. What do you think about that?

**Comment by morazow** [comment]
Hey @jonhoo,

I have added your suggestions.

In other methods: `entry` and `insert`, the compiler did not allow me to use `?` because they do not return an `Option`. I used the `unwrap()` method, I guess it should be safe since we resize if buckets length is zero.

File: src/lib.rs [hunk 1]
```diff
@@ -78,22 +78,25 @@ impl<K, V> HashMap<K, V>
 where
     K: Hash + Eq,
 {
-    fn bucket<Q>(&self, key: &Q) -> usize
+    fn bucket<Q>(&self, key: &Q) -> Option<usize>
     where
         K: Borrow<Q>,
         Q: Hash + Eq + ?Sized,
     {
+        if self.buckets.len() == 0 {
```
Comment (jonhoo) [inline src/lib.rs]: This should probably be `.is_empty()`

File: src/lib.rs [hunk 2]
```diff
@@ -78,22 +78,25 @@ impl<K, V> HashMap<K, V>
...
+        Some((hasher.finish() % self.buckets.len() as u64) as usize)
     }
 
     pub fn entry<'a>(&'a mut self, key: K) -> Entry<'a, K, V> {
         if self.buckets.is_empty() || self.items > 3 * self.buckets.len() / 4 {
             self.resize();
         }
 
-        let bucket = self.bucket(&key);
+        let bucket = self.bucket(&key).unwrap();
```
Comment (jonhoo) [inline src/lib.rs]: Let's make this a `.expect("buckets.is_empty() handled above")`.

File: src/lib.rs (hunk 1 above)
Comment (morazow) [reply src/lib.rs]: Fixed.

File: src/lib.rs (hunk 2 above)
Comment (morazow) [reply src/lib.rs]: :+1:

**Comment by morazow** [comment]
Thanks! I have applied your reviews.

Please let me know if I should squash the commits. Or you can squash merge from the Github interface.

**Review by jonhoo — APPROVED** [review]

**Comment by jonhoo** [comment]
All good, I squashed it :) Thank you!

**Comment by morazow** [comment]
Thank you!

So far it was great experience. I am going to add iterators next.

**Pull request merged**